src/endian/stream_reader.hpp

```diff
@@ -64,11 +64,24 @@ class stream_reader : public stream
     template<class ValueType>
     void read(ValueType& value)
     {
-        // Make sure there is enough data to read in the underlying buffer
-        assert(sizeof(ValueType) <= remaining_size());
+        assert(sizeof(ValueType) <= remaining_size() &&
+               "Reading over the end of the underlying buffer");
+
         read_bytes<sizeof(ValueType), ValueType>(value);
     }
 
+    /// Reads a ValueType-sized integer from the stream and moves the read
+    /// position.
+    ///
+    /// @return the value to be read
+    template<class ValueType>
+    ValueType read()
+    {
+        ValueType value;
```

Comment:
we should probably have the assert here as well:
`assert(sizeof(ValueType) <= remaining_size() && "Reading over the end of the underlying buffer");`